Feat: Hyperlane Event Indexer and Rocks DB

jonas089 opened the pull request:
This PR implements an event indexer that can query all Hyperlane **Dispatch** messages for a given block range (in the test I use 0 to 10000). The **Dispatch** message is emitted from Hyperlane's **Mailbox** contract.

Additionally it implements all utilities to decode Hyperlane Message Headers and Bodies in Rust.

# New Crates Overview

`storage`: RocksDB implementation + simple interface
`evm-state-types`: Types required for Events and Hyperlane Messages, will be extended to also include the types required for the storage proof of the latest branch on-chain (this will be added in a future PR).
`evm-state-queries`: The event indexer itself, that takes an Arc<RwLock<HyperlaneMessageStore>> as input, as well as a websocket url and contract address (default is local testnet, otherwise use .env).

# Environment
New environment variables:

- HYPERLANE_MESSAGE_STORE="testdata/messages/hyperlane"
- RETH_WS_URL
- MAILBOX_CONTRACT_ADDRESS

# Full test setup
## 1. Run docker
`make start`

## 2. Submit transactions
```
make transfer
make transfer-back
```

## 3. run the indexer test
```
cargo test test_run_indexer -- --nocapture
```

The indexer will insert each new **Dispatch** message into the local RocksDB database. A test to decode and print the first message in the local db can be found in `evm-state-queries/src/hyperlane/mod.rs`. 

## Benchmarks and TPS
https://github.com/celestiaorg/hyperlane-tree-bench

damiannolan commented on testdata/messages/hyperlane/000176.sst:
Do we need all of the stuff here in the testdata directory? 

Is it possible to just assign the db path to something like `.db/...` and add to the gitignore file?
If there's specific stuff we need maybe it could be better to have steps to reproduce state rather than checking it in?

jonas089 replied:
I think it's a good idea to change the default path to db/* and add it to .gitignore. Will do that.

damiannolan replied:
nice! thanks

jonas089 commented:
> I tried to run this based on the PR description and ran into an error:
> 
> ```
> cargo run --bin evm-state-listener
> error: no bin target named `evm-state-listener`.
> Available bin targets:
>     evm-exec-program
>     data-gen
>     evm-exec
>     vkey-evm-exec
>     evm-prover
>     evm-range-exec-program
>     evm-range-exec
>     parser
>     vkey-evm-range-exec
> ```
> 
> Maybe these instruction are from a stale commit? I just see library crates in this PR now and no additional binaries.

Yes sorry I just updated the description. I removed the listener in favor of the indexer.

blasrodri commented on crates/evm-state-queries/src/hyperlane/indexer.rs:
```diff
@@ -0,0 +1,124 @@
+        }
+    }
+
+    pub fn from_env() -> Self {
+        dotenvy::dotenv().ok();
+        let socket = WsConnect::new(env::var("RETH_WS_URL").unwrap());
+        let contract_address = Address::from_str(&env::var("MAILBOX_CONTRACT_ADDRESS").unwrap()).unwrap();
+        let filter = Filter::new().address(contract_address).event(&Dispatch::id());
+        Self::new(socket, contract_address, filter)
```
```suggestion
    pub fn from_env() -> Result<Self> {
        dotenvy::dotenv().ok();
        let reth_url = env::var("RETH_WS_URL")
            .map_err(|_| anyhow::anyhow!("RETH_WS_URL environment variable not set"))?;
        let socket = WsConnect::new(reth_url);
        
        let mailbox_addr = env::var("MAILBOX_CONTRACT_ADDRESS")
            .map_err(|_| anyhow::anyhow!("MAILBOX_CONTRACT_ADDRESS environment variable not set"))?;
        let contract_address = Address::from_str(&mailbox_addr)
            .map_err(|e| anyhow::anyhow!("Invalid mailbox contract address: {}", e))?;
            
        let filter = Filter::new().address(contract_address).event(&Dispatch::id());
        Ok(Self::new(socket, contract_address, filter))

blasrodri commented on crates/evm-state-queries/src/hyperlane/indexer.rs:
```diff
@@ -0,0 +1,124 @@
+            match Dispatch::decode_log_data(log.data()) {
+                Ok(event) => {
+                    let dispatch_event: DispatchEvent = event.into();
+                    let current_index = message_store.current_index().unwrap();
```
```suggestion
                    let current_index = message_store.current_index()?;
```

damiannolan commented on crates/storage/src/hyperlane_messages/storage.rs:
```diff
@@ -0,0 +1,101 @@
+}
+
+impl HyperlaneMessageStore {
+    pub fn insert_message(&self, index: u32, message: StoredHyperlaneMessage) -> Result<()> {
+        let read_lock = self.db.read().unwrap();
+        let cf = read_lock.cf_handle("messages").context("Missing CF")?;
+        read_lock.put_cf(cf, index.to_be_bytes(), bincode::serialize(&message)?)?;
+        Ok(())
+    }
```
Is locking required at all? https://github.com/rust-rocksdb/rust-rocksdb?tab=readme-ov-file#multithreaded-columnfamily-alternation

jonas089 commented on crates/storage/src/lib.rs:
```diff
@@ -0,0 +1,14 @@
+pub mod hyperlane_messages;
+
+// every storage module should implement this trait
+pub trait Storage {
```
Yes it's super flexible at the moment, we could get rid of the interface entirely but I thought it'd be good to have one that we can later extend if we see value there.

jonas089 commented on crates/evm-state-types/src/hyperlane.rs:
Yes, I didn't want to pull in their library for just a message type.

jonas089 commented on Cargo.toml:
```diff
@@ -6,10 +6,16 @@ members = [
 ]
 resolver = "2"
 
 [workspace.dependencies]
```
I always run cargo fmt but it doesn't seem to do that.

jonas089 commented:
> If you just run `cargo build` locally and commit the change to the lockfile it should resolve the CI failure

I ran cargo sort --workspace which seems to have fixed it.

jonas089 commented:
@blasrodri Should I remove the storage interface? I'm not sure how useful it'll be moving forward - I could just make this a standalone storage implementation for the Hyperlane message for now (especially since you'll initially have a separate implementation for the proofs).
